### PR TITLE
Make clientId field non-nullable with manual migration

### PR DESCRIPTION
**Pull request by Will-Howard**
The manual migration (`./scripts/serverShellCommand.sh 'Globals.migrations.setNotNullClientIds()'`) needs to be run before this is deployed, the automatic migration will block it from deploying otherwise. @b0b3rt could you run it on LW and let me know when it's done?

---

I [tried before](https://github.com/ForumMagnum/ForumMagnum/pull/9209) to make the clientId field not-nullable and unique, adding the `NOT NULL` constraint ended up locking the table when it was deployed to production, so I had to [revert that part](https://github.com/ForumMagnum/ForumMagnum/pull/9296).

This is a second attempt, using the trick described [here](https://dba.stackexchange.com/questions/267947/how-can-i-set-a-column-to-not-null-without-locking-the-table-during-a-table-scan) to add the constraint without locking the table. I've added this as a manual migration, because it needs to run outside a transaction in order to not lock the table (and it also can't have a transaction waiting on it to finish, as that would cause a deadlock, so it can't use `dbOutsideTransaction` within a regular migration).

The initial goal of that [first PR](https://github.com/ForumMagnum/ForumMagnum/pull/9209) was to use a unique index to do an `INSERT ... ON CONFLICT ("clientId")` query to ensure client ids exist. Not having the `NOT NULL` constraint also broke that because of a quirk of our index query builder (it made it unique by `COALESCE("clientId", '')` rather than `"clientId"`), so I ended up giving up and [working around](https://github.com/ForumMagnum/ForumMagnum/pull/9204/commits/8597a85c4ae5ba12902c7da567f7cf9b9c00f147) the need for a unique index. This PR also fixes that index (sets it to unique by `"clientId"`) so we can undo that workaround (not in this PR).

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207408267599366) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { getSqlClientOrThrow } from "../../lib/sql/sqlClient";
+import { registerMigration } from "./migrationUtils";
+
+export const idxName = 'idx_idx_ClientIds_clientId_unique'
+
+registerMigration({
+  name: "setNotNullClientIds",
```
Comment (jpaddison3) [inline packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts:7]: I take it you've tested this and verified that it doesn't lock?

File: packages/lesswrong/server/migrations/20240526T111308.set_not_null_clientId_2.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import { ClientIds } from "@/lib/collections/clientIds/collection";
+import { updateIndexes } from "./meta/utils";
+import { idxName } from "../manualMigrations/2024-05-26-setNotNullClientIds";
```
Comment (jpaddison3) [inline packages/lesswrong/server/migrations/20240526T111308.set_not_null_clientId_2.ts:3]: I don't love importing from manual migrations here. It violates an expectation that those files are optional for mature dbs and can be deleted if desired.

**Review by jpaddison3 — APPROVED** [review]

File: packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts (hunk 1 above)
Comment (Will-Howard) [reply packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts:7]: Yep

File: packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    // Update index
+    const indexQuery = await sql.oneOrNone(`
+      SELECT indexdef
+      FROM pg_indexes
+      WHERE schemaname = 'public' AND tablename = 'ClientIds' AND indexname = '${idxName}'
+    `)
+    const indexIsCorrect = indexQuery?.indexdef === `CREATE UNIQUE INDEX "${idxName}" ON public."ClientIds" USING btree ("clientId")`;
+
+    if (!indexIsCorrect) {
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts:41]: Oh the blocks are already separate (99% sure)

File: packages/lesswrong/server/migrations/20240526T111308.set_not_null_clientId_2.ts (hunk 2 above)
Comment (Will-Howard) [reply packages/lesswrong/server/migrations/20240526T111308.set_not_null_clientId_2.ts:3]: 👍 Good point I'll just define it in both places

File: packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/server/manualMigrations/2024-05-26-setNotNullClientIds.ts:41]: ? Ain't this nested inside the first?

**Pull request merged**